test/parallel/test-find-package-json.js

```diff
@@ -0,0 +1,142 @@
+'use strict';
+
+const common = require('../common');
+const fixtures = require('../common/fixtures');
+const tmpdir = require('../common/tmpdir');
+const assert = require('node:assert');
+const fs = require('node:fs');
+const { findPackageJSON } = require('node:module');
+const { describe, it } = require('node:test');
+const { pathToFileURL } = require('node:url');
+
+
+describe('findPackageJSON', () => { // Throws when no arguments are provided
+  it('should throw when no arguments are provided', () => {
+    assert.throws(
+      () => findPackageJSON(),
+      { code: 'ERR_INVALID_ARG_TYPE' }
+    );
+  });
+
+  it('should throw when parentURL is invalid', () => {
+    for (const invalid of ['', null, undefined, {}, [], Symbol(), 1, 0]) assert.throws(
+      () => findPackageJSON('', invalid),
+      { code: 'ERR_INVALID_ARG_TYPE' },
+      invalid
+    );
+  });
+
+  it('should accept a file URL (string), like from `import.meta.resolve()`', () => {
+    const importMetaUrl = `${pathToFileURL(fixtures.fixturesDir)}/whatever.ext`;
+    const specifierBase = './packages/root-types-field/';
+    assert.strictEqual(
+      findPackageJSON(`${specifierBase}index.js`, importMetaUrl),
+      fixtures.path(specifierBase, 'package.json')
+    );
+  });
+
+  it('should be able to crawl up (CJS)', () => {
+    const pathToMod = fixtures.path('packages/nested/sub-pkg-cjs/index.js');
+    const parentPkg = require(pathToMod);
+
+    const pathToParent = fixtures.path('packages/nested/package.json');
+    assert.strictEqual(parentPkg, pathToParent);
+  });
+
+  it('should be able to crawl up (ESM)', () => {
+    const pathToMod = fixtures.path('packages/nested/sub-pkg-esm/index.js');
+    const parentPkg = require(pathToMod).default; // This test is a CJS file
+
+    const pathToParent = fixtures.path('packages/nested/package.json');
+    assert.strictEqual(parentPkg, pathToParent);
+  });
+
+  it('can require via package.json', () => {
+    const pathToMod = fixtures.path('packages/cjs-main-no-index/other.js');
+    // `require()` falls back to package.json values like "main" to resolve when there is no index
+    const answer = require(pathToMod);
+
+    assert.strictEqual(answer, 43);
+  });
+
+  it('should be able to do whatever the heck Antoine created', () => {
+    tmpdir.refresh();
+
+    fs.writeFileSync(tmpdir.resolve('entry.mjs'), `
+      import { findPackageJSON, createRequire } from 'node:module';
+      import fs from 'node:fs';
+      const require = createRequire(import.meta.url);
+
+      const readPJSON = (locus) => JSON.parse(fs.readFileSync(locus, 'utf8'));
+
+      const { secretNumber1 } = readPJSON(findPackageJSON('pkg', import.meta.url));
+      // This is impossible to resolve because pkg's pjson.exports blocks access
+      let secretNumber2Location = NaN;
+      try { secretNumber2Location = findPackageJSON('pkg/subfolder/', import.meta.url) } catch {}
```

Comment:
The name of the variable `secretNumber2Location` infers it's a path, yet we set it to `NaN`, that's confusing.
Also, `secretNumber2` is defined in the `package.json` at the root of the `pkg` package, not in the subfolder, so that's extra confusing (my bad for that). I suggest we rename the `secretNumberX` to clearer names, e.g. `secretNumberRoot`, `secretNumberSubfolder`, `secretNumberSubfolder2`, `secretNumberPkg2`.

Reply:
> The name of the variable `secretNumber2Location` infers it's a path, yet we set it to `NaN`, that's confusing.

ah, true. I had left it `undefined`, but then decided that was ambiguous (because undefined could also mean something just never set it that should have). perhaps `null`?



> I suggest we rename the `secretNumberX` to clearer names, e.g. `secretNumberRoot`, `secretNumberSubfolder`, `secretNumberSubfolder2`, `secretNumberPkg2`.

👍 good idea :)